sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -2792,4 +2793,40 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test("`Cast` to CHAR/VARCHAR should truncate the values") {
+    withTable("t") {
+      val m = intercept[ParseException] {
+        sql("SELECT CAST('abc' AS CHAR(0))")
+      }.getMessage
+      assert(m.contains("Char length 0 is out of range [1, 255]"))
+
+      val m2 = intercept[ParseException] {
+        sql("SELECT CAST('abc' AS VARCHAR(0))")
+      }.getMessage
+      assert(m2.contains("VarChar length 0 is out of range [1, 65535]"))
+
+      checkAnswer(
+        sql("SELECT CAST('abc' AS CHAR(2)), CAST('abc' AS CHAR(4))"),
+        Row("ab", "abc"))
+
+      sql("CREATE TABLE t(a STRING) USING PARQUET")
+      sql("INSERT INTO t VALUES ('abc')")
+      sql("INSERT INTO t VALUES (null)")
+
+      checkAnswer(
+        sql("SELECT CAST(a AS CHAR(2)), CAST(a AS CHAR(3)), CAST(a AS CHAR(4)) FROM t"),
+        Row("ab", "abc", "abc") :: Row(null, null, null) :: Nil)
+
+      sql(
+        """
+          |CREATE TABLE t_ctas
+          |USING ORC
+          |AS SELECT CAST(a AS CHAR(2)) c1, CAST(a AS CHAR(3)) c2, CAST(a AS CHAR(4)) c3 FROM t
```

Comment:
I agree that this is partial. Actually, this CTAS test case was I'd like to support in this PR.
Do you think Spark can have `CHAR`/`VARCHAR` in 2.4, @gatorsmile ?

Reply:
Without native support of new data type CHAR and VARCHAR, supporting the syntactic sugar is misleading to end users, in my opinion.

Reply:
Currently, this PR only resolves the top -level CHAR and VARCHAR. When they are embedded, does it still work? To me, the whole solution is hacky.

Reply:
What do you mean by `embedded`? This works in a subquery and in an expression like function arguments. Could you give us some examples?

Reply:
inside another expression.